File: fjlc/utils/reader/LineReader.java
package com.freva.masteroppgave.utils.reader;

import com.freva.masteroppgave.utils.FileUtils;
import com.freva.masteroppgave.utils.progressbar.Progressable;

import java.io.File;
import java.io.IOException;
import java.util.Iterator;
import java.util.Scanner;

public class LineReader implements Iterator<String>, Iterable<String>, Progressable {
    private Scanner scanner;
    private int totalLines = 0;
    private int lineCounter = 0;

    public LineReader(File file) throws IOException {
        this.totalLines = FileUtils.countLines(file);
        this.scanner = new Scanner(file, "UTF-8");
    }

    public boolean hasNext() {
        return scanner.hasNext();
    }

    public String next() {
        lineCounter++;
        return scanner.nextLine();
    }

    public Iterator<String> iterator() {
        return this;
    }

    public double getProgress() {
        return (totalLines == 0 ? 0 : 100.0 * lineCounter / totalLines);
    }
}

File: fjlc/utils/reader/line_reader.py
from fjlc.utils import file_utils


class LineReader:

    def __init__(self, file_name):
        self.total_lines = file_utils.count_lines(file_name)
        self.file = open(file_name)
        self.line_counter = 0

    def has_next(self):
        return self.line_counter < self.total_lines

    def __next__(self):
        if not self.has_next():
            raise StopIteration
        self.line_counter += 1
        return self.file.readline()

    def __iter__(self):
        return self

    def get_progress(self):
        return 0 if self.total_lines else 100.0 * self.line_counter / self.total_lines
